StronglyTypedId: Cannot be used in .NET Framework 4.8

Using version 1.0.33

Code:
``` c#
[StronglyTypedId(typeof(int))]
public partial struct TestId { }
```

I get this error:
```
1>C:\projects\ConsoleApp1\Meziantou.Framework.StronglyTypedId\Meziantou.Framework.StronglyTypedId.StronglyTypedIdSourceGenerator\TestId.g.cs(99,21,99,52): error CS0234: The type or namespace name 'MemoryExtensions' does not exist in the namespace 'System' (are you missing an assembly reference?)
```


![image](https://github.com/meziantou/Meziantou.Framework/assets/249938/a3b5bd3a-1974-483c-8515-c893d3fa459a)

--- a/src/Meziantou.Framework.StronglyTypedId/Meziantou.Framework.StronglyTypedId.csproj
+++ b/src/Meziantou.Framework.StronglyTypedId/Meziantou.Framework.StronglyTypedId.csproj
@@ -3,7 +3,7 @@
     <TargetFrameworks>$(LatestTargetFrameworks);netstandard2.0</TargetFrameworks>
     <CompatibilityRange>true</CompatibilityRange>
     <CompatibilityRecord>true</CompatibilityRecord>
-    <Version>1.0.33</Version>
+    <Version>1.0.34</Version>
 
     <developmentDependency>true</developmentDependency>
     <GeneratePackageOnBuild>false</GeneratePackageOnBuild>

--- a/src/Meziantou.Framework.StronglyTypedId/StronglyTypedIdSourceGenerator.Members.cs
+++ b/src/Meziantou.Framework.StronglyTypedId/StronglyTypedIdSourceGenerator.Members.cs
@@ -161,7 +161,7 @@ public partial class StronglyTypedIdSourceGenerator
                 XmlParam("a", "The first object to compare"),
                 XmlParam("b", "The second object to compare"),
                 XmlReturn(XmlSeeLangword("true"), " if ", XmlParamRef("a"), " and ", XmlParamRef("b"), " are equal; otherwise, ", XmlSeeLangword("false"), "."));
-            writer.WriteLine($"public static bool operator ==({context.CSharpNullableTypeName} a, {context.CSharpNullableTypeName} b) => global::System.Collections.Generic.EqualityComparer<{context.TypeName}>.Default.Equals(a, b);");
+            writer.WriteLine($"public static bool operator ==({context.CSharpNullableTypeName} a, {context.CSharpNullableTypeName} b) => global::System.Collections.Generic.EqualityComparer<{context.CSharpNullableTypeName}>.Default.Equals(a, b);");
         }
 
         // Operator !=
@@ -248,7 +248,7 @@ public partial class StronglyTypedIdSourceGenerator
             }
 
             // Parse(ReadOnlySpan<char>)
-            if (!context.IsParseDefined_Span)
+            if (!context.IsParseDefined_ReadOnlySpan)
             {
                 WriteNewMember(
                     XmlSummary("Converts the read-only character span that represents of a ", XmlSeeCref(context.TypeName), " to the equivalent ", XmlSeeCref(context.TypeName), " type."),
@@ -275,7 +275,17 @@ public partial class StronglyTypedIdSourceGenerator
         {
             using (writer.BeginBlock($"if (TryParse(value, out var result))"))
             {
+                if (!context.SupportNotNullWhenAttribute)
+                {
+                    writer.WriteLine($"#nullable disable");
+                }
+
                 writer.WriteLine($"return result;");
+
+                if (!context.SupportNotNullWhenAttribute)
+                {
+                    writer.WriteLine($"#nullable enable");
+                }
             }
 
             writer.WriteLine($"throw new global::System.FormatException($\"value '{{value}}' is not valid\");");

--- a/src/Meziantou.Framework.StronglyTypedId/StronglyTypedIdSourceGenerator.cs
+++ b/src/Meziantou.Framework.StronglyTypedId/StronglyTypedIdSourceGenerator.cs
@@ -416,7 +416,7 @@ internal sealed class StronglyTypedIdAttribute : System.Attribute
                         break;
 
                     case IMethodSymbol { IsStatic: true, Name: "Parse", Parameters: [var param1, ..] } when SymbolEqual(param1.Type, readOnlySpanCharSymbol):
-                        IsTryParseDefined_ReadOnlySpan = true;
+                        IsParseDefined_ReadOnlySpan = true;
                         break;
 
                     case IMethodSymbol { IsStatic: false, Name: ".ctor", Parameters: [var param1] } when SymbolEqual(param1.Type, idTypeSymbol):
@@ -560,7 +560,7 @@ internal sealed class StronglyTypedIdAttribute : System.Attribute
         public bool IsTryParseDefined_String { get; }
         public bool IsTryParseDefined_ReadOnlySpan { get; }
         public bool IsParseDefined_String { get; }
-        public bool IsParseDefined_Span { get; }
+        public bool IsParseDefined_ReadOnlySpan { get; }
         public bool IsOpLessThanDefined { get; }
         public bool IsOpGreaterThanDefined { get; }
         public bool IsOpLessThanOrEqualDefined { get; }
@@ -593,7 +593,7 @@ internal sealed class StronglyTypedIdAttribute : System.Attribute
         public string CSharpNullableTypeName { get; }
 
         public bool IsValueTypeNullable => IdType is IdType.System_String;
-        public bool ValueTypeHasParseReadOnlySpan => IdType != IdType.MongoDB_Bson_ObjectId;
+        public bool ValueTypeHasParseReadOnlySpan => IdType != IdType.MongoDB_Bson_ObjectId && SupportReadOnlySpanChar;
 
         public override bool Equals(object? obj) => Equals(obj as AttributeInfo);
 
@@ -621,7 +621,7 @@ internal sealed class StronglyTypedIdAttribute : System.Attribute
                 && IsTryParseDefined_String == other.IsTryParseDefined_String
                 && IsTryParseDefined_ReadOnlySpan == other.IsTryParseDefined_ReadOnlySpan
                 && IsParseDefined_String == other.IsParseDefined_String
-                && IsParseDefined_Span == other.IsParseDefined_Span
+                && IsParseDefined_ReadOnlySpan == other.IsParseDefined_ReadOnlySpan
                 && SupportStaticInterfaces == other.SupportStaticInterfaces
                 && SupportIStronglyTyped == other.SupportIStronglyTyped
                 && SupportIStronglyTypedOfT == other.SupportIStronglyTypedOfT
@@ -666,7 +666,7 @@ internal sealed class StronglyTypedIdAttribute : System.Attribute
             hash = (hash * 397) ^ IsTryParseDefined_String.GetHashCode();
             hash = (hash * 397) ^ IsTryParseDefined_ReadOnlySpan.GetHashCode();
             hash = (hash * 397) ^ IsParseDefined_String.GetHashCode();
-            hash = (hash * 397) ^ IsParseDefined_Span.GetHashCode();
+            hash = (hash * 397) ^ IsParseDefined_ReadOnlySpan.GetHashCode();
             hash = (hash * 397) ^ SupportIStronglyTyped.GetHashCode();
             hash = (hash * 397) ^ SupportIStronglyTypedOfT.GetHashCode();
             hash = (hash * 397) ^ SupportStaticInterfaces.GetHashCode();

--- a/tests/Meziantou.Framework.StronglyTypedId.Tests/StronglyTypedIdSourceGeneratorTests.cs
+++ b/tests/Meziantou.Framework.StronglyTypedId.Tests/StronglyTypedIdSourceGeneratorTests.cs
@@ -7,7 +7,6 @@ using System.Text;
 using FluentAssertions;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
-using NuGet.Frameworks;
 using TestUtilities;
 using Xunit;
 
@@ -57,27 +56,11 @@ public sealed class StronglyTypedIdSourceGeneratorTests
         {
             var diags = string.Join("\n", result.Diagnostics);
             var generated = runResult.GeneratedTrees.Length > 1 ? (await runResult.GeneratedTrees[1].GetRootAsync()).ToFullString() : "<no file generated>";
-            result.Success.Should().BeTrue("Project cannot build:\n" + diags + "\n\n\n" + NumberLine(generated));
+            result.Success.Should().BeTrue("Project cannot build:\n" + diags + "\n\n\n" + AddNumberLine(generated));
             result.Diagnostics.Should().BeEmpty();
         }
 
         return (runResult, outputCompilation, result.Success ? ms.ToArray() : null);
-
-        static string NumberLine(string value)
-        {
-            var sb = new StringBuilder();
-            var i = 0;
-            foreach (var line in value.SplitLines())
-            {
-                i++;
-                sb.Append(i.ToStringInvariant("#000"));
-                sb.Append(' ');
-                sb.Append(line.Line);
-                sb.Append(line.Separator);
-            }
-
-            return sb.ToString();
-        }
     }
 
     [Fact]
@@ -420,7 +403,7 @@ public partial struct Test : System.IComparable<Test> {}
         result.GeneratorResult.Diagnostics.Should().BeEmpty();
         result.GeneratorResult.GeneratedTrees.Should().HaveCount(2);
     }
-    
+
     [Fact]
     public async Task Generate_IComparable_Class_ReferenceType()
     {
@@ -540,7 +523,26 @@ public partial class Test : System.IComparable<Test> {}
 
         static IEnumerable<BuildMatrixArguments> BuildMatrixTestCases()
         {
-            var types = new[] { "string", "int" };
+            var types = new[]
+            {
+                "System.Boolean",
+                "System.Byte",
+                "System.DateTime",
+                "System.DateTimeOffset",
+                "System.Decimal",
+                "System.Double",
+                "System.Guid",
+                "System.Int16",
+                "System.Int32",
+                "System.Int64",
+                "System.SByte",
+                "System.Single",
+                "System.String",
+                "System.UInt16",
+                "System.UInt32",
+                "System.UInt64",
+            };
+
             var declarations = new[]
             {
                 "partial struct Test { }",
@@ -557,6 +559,14 @@ public partial class Test : System.IComparable<Test> {}
             {
                 foreach (var declaration in declarations)
                 {
+                    foreach (var netFrameworkVersion in new[] { "462", "472", "481" })
+                    {
+                        yield return new BuildMatrixArguments(type, declaration, new[]
+                        {
+                            new NuGetReference("Microsoft.NETFramework.ReferenceAssemblies.net" + netFrameworkVersion, "1.0.3", ""),
+                        });
+                    }
+
                     foreach (var netcoreVersion in new[] { "5.0.0", "6.0.12", "7.0.1" })
                     {
                         yield return new BuildMatrixArguments(type, declaration, new[] { new NuGetReference("Microsoft.NETCore.App.Ref", netcoreVersion, "ref/") });
@@ -579,6 +589,24 @@ public partial class Test : System.IComparable<Test> {}
                             new NuGetReference("MongoDB.Bson", mongodb, "lib/netstandard2.1/"),
                         });
                     }
+                }
+            }
+
+            // Add specific test cases
+            foreach (var declaration in declarations)
+            {
+                yield return new BuildMatrixArguments("System.Half", declaration, new[] { new NuGetReference("Microsoft.NETCore.App.Ref", "7.0.1", "ref/") });
+                yield return new BuildMatrixArguments("System.Int128", declaration, new[] { new NuGetReference("Microsoft.NETCore.App.Ref", "7.0.1", "ref/") });
+                yield return new BuildMatrixArguments("System.UInt128", declaration, new[] { new NuGetReference("Microsoft.NETCore.App.Ref", "7.0.1", "ref/") });
+                yield return new BuildMatrixArguments("System.Numerics.BigInteger", declaration, new[] { new NuGetReference("Microsoft.NETCore.App.Ref", "7.0.1", "ref/") });
+
+                foreach (var mongodb in new[] { "2.18.0" })
+                {
+                    yield return new BuildMatrixArguments("MongoDB.Bson.ObjectId", declaration, new[]
+                    {
+                            new NuGetReference("Microsoft.NETCore.App.Ref", "7.0.1", "ref/"),
+                            new NuGetReference("MongoDB.Bson", mongodb, "lib/netstandard2.1/"),
+                        });
                 }
             }
         }
@@ -607,7 +635,23 @@ public partial class Test : System.IComparable<Test> {}
 
         var diags = string.Join("\n", compilationOutput.Diagnostics);
         var generated = runResult.GeneratedTrees.Length > 1 ? (await runResult.GeneratedTrees[1].GetRootAsync()).ToFullString() : "<no file generated>";
-        compilationOutput.Success.Should().BeTrue("Project cannot build:\n" + diags + "\n\n\n" + generated);
+        compilationOutput.Success.Should().BeTrue("Project cannot build:\n" + diags + "\n\n\n" + AddNumberLine(generated));
         compilationOutput.Diagnostics.Should().BeEmpty();
+    }
+
+    private static string AddNumberLine(string value)
+    {
+        var sb = new StringBuilder();
+        var i = 0;
+        foreach (var line in value.SplitLines())
+        {
+            i++;
+            sb.Append(i.ToStringInvariant("#000"));
+            sb.Append(' ');
+            sb.Append(line.Line);
+            sb.Append(line.Separator);
+        }
+
+        return sb.ToString();
     }
 }

